sources/SilkTouch/SilkTouch/Mods/TransformCOM.cs

```diff
@@ -0,0 +1,749 @@
+                {
+                    var variableExpression = variables[i].Initializer?.Value as LiteralExpressionSyntax;
+                    // Check if the variable is instantiated to a null literal
+                    if (variableExpression is not null && variableExpression.IsKind(SyntaxKind.NullLiteralExpression))
+                    {
+                        var newDefaultArg = LiteralExpression(SyntaxKind.DefaultLiteralExpression);
+                        newInstantiation = newInstantiation.ReplaceNode(variableExpression, newDefaultArg);
+                    }
+                }
+                return newInstantiation;
+            }
+
+            private string GetTypeName(string fullTypeName)
+            {
+                int dot = fullTypeName.LastIndexOf('.');
+
+                if (dot != -1)
+                {
+                    return fullTypeName.Substring(dot + 1);
+                }
+
+                return fullTypeName;
+            }
+        }
+
+        class Rewriter(List<(string, bool, KeyedStringTree?)> ComTypes)
+            : CSharpSyntaxRewriter
+        {
+            public override SyntaxNode? VisitPointerType(PointerTypeSyntax node)
+            {
+                for (int i = 0; i < ComTypes.Count; i++)
+                {
+                    (string, bool, KeyedStringTree?) val = ComTypes[i];
+
+                    if (val.Item1 == node.ElementType.ToString())
+                    {
+                        return IdentifierName(val.Item1);
+                    }
+                }
+
+                return base.VisitPointerType(node);
+            }
+
+            public override SyntaxNode? VisitIdentifierName(IdentifierNameSyntax node)
+            {
+                if (node.Identifier.ToString() == "lpVtbl")
+                {
+                    return ParenthesizedExpression(PrefixUnaryExpression(SyntaxKind.PointerIndirectionExpression, node));
+                }
+
+                return base.VisitIdentifierName(node);
+            }
+
+            public override SyntaxNode? VisitVariableDeclaration(VariableDeclarationSyntax node)
+            {
+                if (node.Type.ToString() == "void**" && node.Variables.First().Identifier.ToString() == "lpVtbl")
+                {
+                    return VariableDeclaration(PointerType(PointerType(PointerType(IdentifierName("void")))))
+                        .AddVariables(VariableDeclarator("lpVtbl"));
+                }
+
+                return base.VisitVariableDeclaration(node);
+            }
+
+            public override SyntaxNode? VisitInterfaceDeclaration(InterfaceDeclarationSyntax node)
+            {
+                var ret = base.VisitInterfaceDeclaration(node);
+
+                if (ret is InterfaceDeclarationSyntax inter && inter.BaseList is not null && inter.BaseList.Types.Any(baseType => baseType.Type.ToString().StartsWith("I") && baseType.Type.ToString().EndsWith(".Interface")))
```

Comment:
`StarsWith("I")` feels sus, is there a better check we can do here? Thinking about WinRT where some types using the COM ABI do not start with I.

Reply:
probably fair, cleaning up

Reply:
You are probably right, I'll clean this up

Reply:
fixed